aur-fetch: miscellaneous fixes

AladW opened the pull request:
Fixing some more breakage. In particular #636 surfaced again.

AladW commented:
Playing around with the new `aur-fetch` and `aur-sync`, I've noticed that cloning new repositories, and then existing `vifm` with `:cq` resulted in a `.diff` file for _every_ cloned directory on the following `aur-sync` invocation. The issue is that `AUR_SEEN` is not set due to exiting early, after which aur-fetch falls back to the empty tree object for diffs.

I see two options to handle this:

1. The aforementioned (#652) fallback to `HEAD`, instead of the empty tree object. (`HEAD` can be used directly instead of `ORIG_HEAD`, by placing the `seen` check above the `reset` resp. `rebase` steps).

2. Have `aur-sync` unconditionally update `AUR_SEEN` to `HEAD` for any newly cloned repositories. This can be achieved with `aur-fetch --results`.
   
Some other things I noticed while thinking about this issue:

* `--confirm-seen` is part of the main loop. In particular, dependencies from `--recurse` may be also be marked as seen if this flag is enabled, and any cloned repositories directly after they are cloned.
  
  This may be problematic when the respective packages (and their dependencies) change between the time the user was confident in their contents, and the time he runs `aur fetch --confirm-seen`.
  + Suggested change: move `--confirm-seen` outside the main loop, and only apply it to command-line arguments.

* To support `AUR_SEEN` and the according "persistent" diffs with `aur-sync --continue`, `aur-fetch` should have the ability to not run `aur-fetch` (and any subsequent `rebase` or `reset` steps).
  + Suggested change: add an `--existing` flag to `aur-fetch`, which groups `git fetch -v`, the `rebase`/`reset` logic, and `results` together in an `if` list. The condition in this list should be true by default.

* The `clone` step should likely initialize `AUR_SEEN` to a valid value (i.e. the empty tree object), regardless of `--confirm-seen`. The user can then set it to the actual value (i.e. `HEAD`) with `--confirm-seen`. In particular this would avoid `aur-fetch` not finding `AUR_SEEN` in its own repositories (and add pointless warnings accordingly).
  + Suggested change: update `AUR_SEEN` to the empty tree object on
    the `git clone` step in `aur-fetch`.

* In `2.3`, `aur-fetch` would only run `git reset` if `HEAD` differs from `HEAD@{upstream}`. In particular, the user could make ad-hoc changes to the git repositories (e.g. the common missing `makedepends` for `chroot` builds), which would at least be preserved until the next upstream commit. With the current revision, `git reset` is run on _every run_ of `aur-sync`, which is a far more aggressive approach that makes small changes harder (`git commit`, set `aurutils.rebase`, hope `rebase` applies on the next upstream commit, etc.)
  
  This is not a general stab at `rebase`, but more that it's not the solution to everything and as such shouldn't be imposed for every (small) kind of modification.
  + Suggested change: update `aur-fetch -S` resp. `--sync=auto` to check for upstream commits, e.g. by comparing `FETCH_HEAD` before and after `git fetch` is run. `--sync=rebase` and `--sync=reset` would still `rebase` and `reset` unconditionally (and could be exposed to `aur-sync`, if desired).

* Really a nitpick; when investigating the issues above, I found the code fairly hard to navigate with all details of `git` commands exposed in the main loop.
  + Suggested change: add some wrapper functions with descriptive names, e.g. `aur_rev` for `git rev-parse --quiet --verify` and `aur_reset` for `git reset --hard 'HEAD@{upstream}'`.

AladW commented:
Addendum: Comparing `FETCH_HEAD` is not a great solution, since running `git fetch` outside of `aur fetch` would prevent `git rebase` or `git reset` from being run until the next upstream revision. The following could work instead:

```
if [[ $(git config --get --type=bool aurutils.rebase) == 'true' ]]; then
  plain >&2 'aurutils.rebase is set for %s' "$pkg"

  if ! git branch --contains 'HEAD@{upstream}'; then
    aur_rebase
  fi
else
  if [[ $(aur_rev 'HEAD') != $(aur_rev 'HEAD@{upstream}') ]]; then
    aur_reset
  fi
fi
```

rafasc commented:
> Playing around with the new `aur-fetch` and `aur-sync`, I've noticed that cloning new repositories, and then existing `vifm` with `:cq` resulted in a `.diff` file for _every_ cloned directory on the following `aur-sync` invocation. The issue is that `AUR_SEEN` is not set due to exiting early, after which aur-fetch falls back to the empty tree object for diffs.
> 
> I see two options to handle this:
> 
> 1. The aforementioned (#652) fallback to `HEAD`, instead of the empty tree object. (`HEAD` can be used directly instead of `ORIG_HEAD`, by placing the `seen` check above the `reset` resp. `rebase` steps).
> 2. Have `aur-sync` unconditionally update `AUR_SEEN` to `HEAD` for any newly cloned repositories. This can be achieved with `aur-fetch --results`.
> 
I don't agree with this. This would make `aur-fetch`, and indirectly `aur-sync`, *trust-by-default* in certain situations; some of them beyond our control (like using repos that were not cloned by us) etc.  Not to mention it would make `:cq` in `aur-sync` inconsistent as it would sometimes mean abort the transaction but other times something else.

> Some other things I noticed while thinking about this issue:
> 
> * `--confirm-seen` is part of the main loop. In particular, dependencies from `--recurse` may be also be marked as seen if this flag is enabled, and any cloned repositories directly after they are cloned.
>   This may be problematic when the respective packages (and their dependencies) change between the time the user was confident in their contents, and the time he runs `aur fetch --confirm-seen`.
>   
>   * **Suggested change:** move `--confirm-seen` outside the main loop, and only apply it to command-line arguments.

The `--confirm-seen` mode skips the fetch and any rebase/reset operations. It just marks the checked out branch as trusted. You're not confirming what's on the server, you're confirming what's on the local machine.

That being said, there is a small flaw in the aur-fetch --recursive, because the output of `aur-depends` is not guaranteed to be the same between the invocations of `aur-fetch --recursive X` and `aur -fetch --recursive --confirm-seen X`.

If we want to cover that we would need something way more complex than this. (As a side note, `aur-sync` is not affected by this, because we store the results of aur depends and perform each fetch individually, i.e. we do not use `--recurse`.)

Also, with the above suggested change it would treat dependencies fetched via --recursive  *trusted-by-default*, again nothing something we should be doing.

Any of the above would be enough to change `File view` and/or `Diff view` to Partial in the wiki.


> * To support `AUR_SEEN` and the according "persistent" diffs with `aur-sync --continue`, `aur-fetch` should have the ability to not run `aur-fetch` (and any subsequent `rebase` or `reset` steps).
>   
>   * **Suggested change:** add an `--existing` flag to `aur-fetch`, which groups `git fetch -v`, the `rebase`/`reset` logic, and `results` together in an `if` list. The condition in this list should be true by default.

The only thing missing in your suggested change is that it also should skip cloning new packages. But otherwise shouldn't be hard to implement. :+1: 


> * The `clone` step should likely initialize `AUR_SEEN` to a valid value (i.e. the empty tree object), regardless of `--confirm-seen`. The user can then set it to the actual value (i.e. `HEAD`) with `- confirm-seen`. In particular this would avoid `aur-fetch` not finding `AUR_SEEN` in its own repositories (and add pointless warnings accordingly).
>   
>   * **Suggested change:** update `AUR_SEEN` to the empty tree object on the `git clone` step in `aur-fetch`.

Also :+1:

This is likely the easy way to deal with 1) / 2). I.e. When aur-fetch is actually in the `fetch` part of the loop: 
- Error if `AUR_SEEN` is missing (This is something that was not cloned by us)
- Notify via results_file if `AUR_SEEN` is set to the empty tree ¹
- Continue otherwise

<sup>¹  we may need to change the `clone|fetch|error` categories to accommodate the meaning, or just mark it as `clone` as it acts as one from the users point of view.</sup>

> * In `2.3`, `aur-fetch` would only run `git reset` if `HEAD` differs from `HEAD@{upstream}`. In particular, the user could make ad-hoc changes to the git repositories (e.g. the common missing `makedepends` for `chroot` builds), which would at least be preserved until the next upstream commit. With the current revision, `git reset` is run on _every run_ of `aur-sync`, which is a far more aggressive approach that makes small changes harder (`git commit`, set `aurutils.rebase`, hope `rebase` applies on the next upstream commit, etc.)
>   This is not a general stab at `rebase`, but more that it's not the solution to everything and as such shouldn't be imposed for every (small) kind of modification.
>  

Your https://github.com/AladW/aurutils/pull/659#issuecomment-570203841 should work unless I am missing something, but the rebase condition should be: `if ! git rev-parse --is-ancestor HEAD@{upstream} HEAD`

AladW commented:
> Any of the above would be enough to change File view and/or Diff view to Partial in the wiki.

I don't understand the argument. The proposed changes don't remove anything from what's been in 2.3 and previous versions for years. It only adds "persistent" diffs, while removing the possibility to have `/dev/null` diffs of questionable value.

> That being said, there is a small flaw in the aur-fetch --recursive, because the output of aur-depends is not guaranteed to be the same between the invocations of aur-fetch --recursive X and aur -fetch --recursive --confirm-seen X.

Yes, that's precisely the point. `--confirm-seen` should only explicitly apply to what's specified on the command-line, not do magic behind the scenes when a repository happens to be cloned, or applied to invisible/implicit arguments from `aur depends`.

> The only thing missing in your suggested change is that it also should skip cloning new packages. But otherwise shouldn't be hard to implement.

Yes. I wasn't sure how to implement this, so I left it out for now.

> This is likely the easy way to deal with 1) / 2).

I guess this works. Though the more I look at it, the less appealing the `AUR_SEEN` thing becomes to me, in particular for a new release.

AladW commented:
Another option is to rename `AUR_SEEN` to `AUR_LAST_DIFF`, to change the intent from "trust" to a less weighed "this is the diff I have shown last time".

AladW commented:
On IRC it was mentioned that the scope of aur-fetch grew too large. If I recall the `update-ref` stuff was first proposed to be added to `aur-sync`; maybe we can revisit this option.

rafasc commented:
I think initially we were going for something very simple:
``` diff 
-xargs -a "$tmp/queue" aur fetch --confirm-seen
+xargs -a "$tmp/queue" -I% git -C % update-ref AUR_SEEN HEAD
```
But if I remember correctly, we decided to put it into `aur-fetch` to take advantage of the checks we already perform there to ensure we only run git commands on valid repos.

On the branch I am experimenting with, I moved it into its own command that we could name `aur-cache` or `aur-review` if we find that are other operations concerning cache management/review worth of supporting by default. (e.g `aur-gc`)

But as of now it only does `--confirm-seen`, so I'm not sure it deserves its own subcommand for such a simple task. Maybe we can do the simple version in `aur-sync` and leave the `aur-cache/review` for `PATH` scripts for now? :thinking:

AladW commented:
> But if I remember correctly, we decided to put it into aur-fetch to take advantage of the checks we already perform there to ensure we only run git commands on valid repos.

This shouldn't be an issue anymore, since we dropped `tar` support for `aur-fetch`, added `--results` and exit directly on errors. (At a later point, we could read `error:` from `--results` instead.)

I made some changes https://github.com/AladW/aurutils/commit/871dedf5b90d13901c1926be6519a73c7ba85395. Purely conceptual since `:cq` on newly cloned directories will cause `aur-sync` to always fail, unless `AUR_SEEN` is set manually.

AladW commented:
@rafasc Any objections against the two commits in this PR? In light of #675 I'd merge them and continue this discussion at another time.

AladW commented:
Considering I have an entirely different approach for `aur-fetch` and `aur-sync` in mind, using a directory structure with "seen" commits instead of an explicit `AUR_SEEN` reference, I'm closing this PR.

AladW commented on lib/aur-fetch:
```diff
@@ -169,14 +167,14 @@ fi | while read -r pkg; do
+        exit 22
 
+    # Otherwise, try to clone anew
+    elif git clone "$AUR_LOCATION/$pkg" "$GIT_DIR"; then
```
Should be fixed in #675.

AladW commented on lib/aur-fetch:
```diff
@@ -112,15 +112,13 @@ if (( recurse )); then
+    export GIT_DIR=$pkg/.git GIT_WORK_TREE=$pkg
 
+    # Reset/rebase if we are on valid repo
+    if [[ -d $pkg ]] && git rev-parse --git-dir >/dev/null; then
```
Thanks for the analysis. My current approach is to check `[[ -d $pkg/.git ]]` as in 2.3, and let `git fetch` fail naturally if the git repository is malformed.
